### DISCUSSION TITLE
Improve grid mode width computation

### DISCUSSION THREAD

**Discussion by akashpal-21** [minad/vertico#415] [Ideas]
Hello, I am having this problem that in grid mode, certain characters are being truncated wrongfully when the search is being successively refined. i do not know how to describe this, so I am attaching a video to clarify this problem, please advice how to solve this.

Only happens when "Indexed and Grid" used together

https://github.com/minad/vertico/assets/46517170/b96925c6-0e4d-428a-b898-b03296a4cbb2

Following is all the configuration I have done,

```

;; 1.4   -  Vertico Completions Settings
(vertico-mode)                                                                  ; substitute vanilla completing read with vertico
(require 'vertico-truncate)                                                     ; enable support for long lines in consult-ripgrep
(vertico-truncate-mode 1)

;;; When resizing the minibuffer (e.g., via the mouse),
;;; adjust the number of visible candidates in Vertico automatically.
;;; n.b.: resize-mini-windows 'grow-only
;;  see : https://github.com/minad/vertico/issues/5
(defun vertico-resize--minibuffer ()
  (add-hook 'window-size-change-functions
	    (lambda (win)
	      (let ((height (window-height win)))
		(when (/= (1- height) vertico-count)
		  (setq-local vertico-count (1- height))
		  (vertico--exhibit))))
	    t t))

(advice-add #'vertico--setup :before #'vertico-resize--minibuffer)
;;;


;; Update minibuffer history with candidate insertions
;; Normally history is updated after minibuffer is closed
;; overrides def. behaviour to update history dynamically
(defadvice vertico-insert
    (after vertico-insert-add-history activate)
  "Make vertico-insert add to the minibuffer history."
  (unless (eq minibuffer-history-variable t)
    (add-to-history minibuffer-history-variable (minibuffer-contents))))
;;;


;; 1.4.1 *  Vertico UI Settings
(vertico-multiform-mode 1)
;; reverse and mouse dont work together
;;(vertico-mouse-mode 1)

(setq vertico-multiform-commands
      '(
	(execute-extended-command reverse indexed grid)
	(consult-buffer reverse indexed)
	(switch-to-buffer reverse indexed grid unobtrusive)
	(kill-buffer reverse indexed)
	(org-roam-node-find reverse grid indexed)
	(org-roam-search reverse indexed)
	(t reverse indexed grid)                                                     ; default for everything else
	))
;;;

```

p.s. i tried disabling vertico-truncate-mode, but this doesn't seem to be related to that

**Comment by minad** [comment]
Thanks for the report. The problem is that the width computation for the grid does not take into account candidate formatting and annotations. This is mainly done for simplicity and performance. See here:

https://github.com/minad/vertico/blob/017ff44443bad401097b9987849625cafa348d86/extensions/vertico-grid.el#L122-L127

I consider this a minor display issue. We could use a different approach where the candidates are annotated first, formatted and then the length and alignment is computed. It will be a little more costly, but if we don't look too far back and ahead it should work well. Patches implementing this are welcome. Converted the issue to an ideas discussion.
